### PR TITLE
Replace ArrayDeque::pollFirst with ArrayDeque::pollLast for Recycler.LocalPool batch. (#14268)

### LINKED ISSUE #14268: Recycler could use a LIFO local structure instead of an ArrayDeque
Currently the Recycler is using an `ArrayDeque` to handle single-same-threaded allocations, see https://github.com/netty/netty/blob/netty-4.1.112.Final/common/src/main/java/io/netty/util/Recycler.java#L300

In the ideal scenario of a pool which has already N elements in (recycled afer a burst of usage - yet to verify how much realistic can be this), if a new claim and release happen on a pooled element, the release will move the released element by the end of the `ArrayDeque`, meaning that a subsequent claim *won't* pick it again, but the next in line recycled element instead, wasting some precious cache locality effect.

Turning the `ArrayDeque` in an `ArrayList` used as a stack can be a better option, but require some benchmarking to verify that is a valid improvement.

### PR DESCRIPTION
**Pull request by Obolrom**
Motivation:

Improve performance by replacing ArrayDeque::pollFirst with ArrayDeque::pollLast.

Modifications:

Replaced ArrayDeque::pollFirst with ArrayDeque::pollLast for LocalPool batch.

Result:
- The new version shows the best overall performance in benchmarks such as plainNew, recyclerGetAndRecycle, and unguardedProducerConsumer.
- Overall, the new version shows a significant reduction in GC time, improving memory efficiency across most tests.

Fixes #14268

### CODE REVIEW

File: common/src/main/java/io/netty/util/Recycler.java [hunk 1]
```diff
@@ -325,7 +326,7 @@ DefaultHandle<T> claim() {
             if (batch.isEmpty()) {
                 handles.drain(this, chunkSize);
             }
-            DefaultHandle<T> handle = batch.pollFirst();
+            DefaultHandle<T> handle = batch.isEmpty() ? null : batch.remove(0);
```
Comment (franz1981) [inline common/src/main/java/io/netty/util/Recycler.java:329]: To use is a queue, it should have been batch.remove(batch.size() - 1)

Said that, maybe just using ArrayDeque::pollLast could be enough, looking at the implementation in the jdk repository

**Review by franz1981 — CHANGES_REQUESTED** [review]

**Comment by franz1981** [comment]
Many thanks for your interest @Obolrom !
Did you sign already our ICLA?

File: common/src/main/java/io/netty/util/Recycler.java [hunk 2]
```diff
@@ -306,7 +307,7 @@ private static final class LocalPool<T> implements MessagePassingQueue.Consumer<
         LocalPool(int maxCapacity, int ratioInterval, int chunkSize) {
             this.ratioInterval = ratioInterval;
             this.chunkSize = chunkSize;
-            batch = new ArrayDeque<DefaultHandle<T>>(chunkSize);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/Recycler.java:309]: yeah I think we should just use `ArrayDeque` ...

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Please make the suggested changes

**Comment by Obolrom** [comment]
@franz1981 
Thanks for the feedback. I will take a look today.
Regarding ICLA, I have signed it.

**Comment by Obolrom** [comment]
## Benchmark Results and Analysis

I have conducted benchmarks for the cases:
1. **Baseline (`ArrayDeque::pollFirst`)**
2. **`ArrayDeque::pollLast`**
3. **ArrayList `batch.remove(0)`**
4. **ArrayList `batch.remove(batch.size() - 1)`**

### Benchmark Comparison

Here’s a detailed comparison of the performance and GC metrics for each of the test cases across different benchmarks:

### 1. `plainNew`
- **Baseline (`ArrayDeque::pollFirst`):** `11.526 ± 0.278 ns/op`
- **`ArrayDeque::pollLast`:** `11.433 ± 0.256 ns/op`
- **ArrayList `batch.remove(0)`:** `11.638 ± 0.189 ns/op`
- **ArrayList `batch.remove(batch.size() - 1)`:** `11.514 ± 0.167 ns/op`

**`ArrayDeque::pollLast`** is the fastest in this test, with a score of `11.433 ns/op`, slightly outperforming the baseline and other versions. Additionally, it exhibits improved GC metrics:
- **GC Time:**
  - **Baseline:** `423 ms`
  - **`ArrayDeque::pollLast`:** `398 ms`

### 2. `producerConsumer`
- **Baseline (`ArrayDeque::pollFirst`):** `125.163 ± 4.603 ns/op`
- **`ArrayDeque::pollLast`:** `125.703 ± 4.695 ns/op`
- **ArrayList `batch.remove(0)`:** `119.510 ± 11.210 ns/op`
- **ArrayList `batch.remove(batch.size() - 1)`:** `144.694 ± 3.844 ns/op`

Although **ArrayList `batch.remove(0)`** is the fastest here (`119.510 ns/op`), **`ArrayDeque::pollLast`** maintains a better balance between performance and memory consumption:
- **GC Alloc Rate:**
  - **Baseline:** `1.109 MB/sec`
  - **`ArrayDeque::pollLast`:** `1.230 MB/sec`
  - **ArrayList `batch.remove(0)`:** `12.534 MB/sec` (significantly higher, indicating more memory usage)

### 3. `recyclerGetAndOrphan`
- **Baseline (`ArrayDeque::pollFirst`):** `14.935 ± 0.401 ns/op`
- **`ArrayDeque::pollLast`:** `15.136 ± 0.573 ns/op`
- **ArrayList `batch.remove(0)`:** `13.825 ± 0.219 ns/op`
- **ArrayList `batch.remove(batch.size() - 1)`:** `13.662 ± 0.262 ns/op`

Both ArrayList methods (`batch.remove(0)` and `batch.remove(batch.size() - 1)`) perform slightly better in terms of execution time, but **`ArrayDeque::pollLast`** still shows strong memory management:
- **GC Alloc Rate:**
  - **Baseline:** `5303 MB/sec`
  - **`ArrayDeque::pollLast`:** `5237 MB/sec`
  - **ArrayList `batch.remove(0)`:** `5725 MB/sec`

**`ArrayDeque::pollLast`** also reduces GC time to `316 ms`, compared to the baseline’s `332 ms`.

### 4. `recyclerGetAndRecycle`
- **Baseline (`ArrayDeque::pollFirst`):** `15.349 ± 0.134 ns/op`
- **`ArrayDeque::pollLast`:** `15.296 ± 0.076 ns/op`
- **ArrayList `batch.remove(0)`:** `16.236 ± 0.773 ns/op`
- **ArrayList `batch.remove(batch.size() - 1)`:** `17.525 ± 1.417 ns/op`

**`ArrayDeque::pollLast`** outperforms all other versions with the best execution time of `15.296 ns/op`, and it does so with minimal GC overhead.

### 5. `unguardedProducerConsumer`
- **Baseline (`ArrayDeque::pollFirst`):** `98.373 ± 12.876 ns/op`
- **`ArrayDeque::pollLast`:** `80.781 ± 14.566 ns/op`
- **ArrayList `batch.remove(0)`:** `101.180 ± 7.570 ns/op`
- **ArrayList `batch.remove(batch.size() - 1)`:** `113.376 ± 7.215 ns/op`

In this benchmark, **`ArrayDeque::pollLast`** demonstrates a significant improvement, outperforming the baseline and other methods by a wide margin (`80.781 ns/op`):
- **GC Alloc Rate:**
  - **Baseline:** `4.808 MB/sec`
  - **`ArrayDeque::pollLast`:** `1.315 MB/sec`

It makes **`ArrayDeque::pollLast`** the most efficient in terms of performance but in terms of memory usage as well.

---
### TLDR; Conclusion

The **`ArrayDeque::pollLast`** version offers a balanced solution across all benchmarks, providing:
- The best overall performance in benchmarks such as `plainNew`, `recyclerGetAndRecycle`, and `unguardedProducerConsumer`.
- A significant reduction in GC time, improving memory efficiency across most tests.
  
It seems **`ArrayDeque::pollLast`** is the optimal choice due to its balanced approach to both performance and resource management.

@franz1981 @normanmaurer what do you think?

**Comment by franz1981** [comment]
I'm curious about

```
5. unguardedProducerConsumer
Baseline (ArrayDeque::pollFirst): 98.373 ± 12.876 ns/op
ArrayDeque::pollLast: 80.781 ± 14.566 ns/op
```
can you verify it more e.g.

by running `-prof perfnorm` (if you're on linux) and publishing the counters there.

Consider that the case where it should grant benefit(s) is with this process:
1. N buffers allocated (and not released)
2. N buffers released (the ones allocated before)
3. 1 keep on been acquired and released 

1 and 2 should be performed in the setup phase and are meant to populate the local queue of N entries.
3 should 
- with `pollFirst`: acquiring from the head, and releasing to the tail - basically iterating the previously populated entries in the local pool
- with `pollLast`: acquiring and releasing the head -> which means reusing the same buffer over and over

**Comment by Obolrom** [comment]
@franz1981
Regarding your last comment, I’m not sure I fully understand what needs to be done.

Do we need to configure the `RecyclerBenchmark` this way?
```java
protected ChainedOptionsBuilder newOptionsBuilder() throws Exception {
     return super.newOptionsBuilder().addProfiler("gc").addProfiler(LinuxPerfNormProfiler.class);
}
```
It provides a huge output with about 150 lines.

Or are we supposed to write a new benchmark from scratch?

**Review by chrisvest — APPROVED** [review]
+1 on using `ArrayDeque::pollLast`. Algorithmically it makes the most sense.
Please update the PR description to match the changes.

**Comment by normanmaurer** [comment]
@franz1981 please pull in once satisfied .

**Comment by franz1981** [comment]
@Obolrom no worries for the benchmarks - in this case can be done later, because TBH the change is just the "right thing" - as @chrisvest said ;) well done!!

**Review by franz1981 — APPROVED** [review]
LGTM and well done picking the issue :)

cannot wait for more!

**Review by normanmaurer — APPROVED** [review]
Maybe just consider adding a comment why we use `pollLast()`
